modules/livestore/partition_reader.go

```diff
@@ -0,0 +1,307 @@
+	records := make([]record, 0, len(fetches.Records()))
+
+	var (
+		minOffset  = int64(math.MaxInt64)
+		maxOffset  = int64(0)
+		totalBytes = int64(0)
+		lastOffset = int64(0)
+	)
+	fetches.EachRecord(func(rec *kgo.Record) {
+		minOffset = min(minOffset, rec.Offset)
+		maxOffset = max(maxOffset, rec.Offset)
+		totalBytes += int64(len(rec.Value))
+		records = append(records, record{
+			content:  rec.Value,
+			tenantID: string(rec.Key),
+			offset:   rec.Offset,
+		})
+		lastOffset = max(lastOffset, rec.Offset)
+	})
+
+	r.highWatermark.Swap(lastOffset)
+
+	// Pass offset and byte information to the live-store
+	err := r.consume(ctx, records)
+	if err != nil {
+		level.Error(r.logger).Log("msg", "encountered error processing records; skipping", "min_offset", minOffset, "max_offset", maxOffset, "err", err)
+		// TODO abort ingesting & back off if it's a server error, ignore error if it's a client error
+	}
+}
+
+func (r *PartitionReader) recordFetchesMetrics(fetches kgo.Fetches) {
+	var (
+		now        = time.Now()
+		numRecords = 0
+	)
+
+	fetches.EachRecord(func(record *kgo.Record) {
+		numRecords++
+		r.metrics.receiveDelay.Observe(now.Sub(record.Timestamp).Seconds())
+	})
+
+	r.metrics.recordsPerFetch.Observe(float64(numRecords))
+}
+
+func (r *PartitionReader) fetchLastCommittedOffsetWithRetries(ctx context.Context) (offset kgo.Offset, err error) {
+	retry := backoff.New(ctx, backoff.Config{
+		MinBackoff: 100 * time.Millisecond,
+		MaxBackoff: 2 * time.Second,
+		MaxRetries: 10,
+	})
+
+	for retry.Ongoing() {
+		offset, err = r.fetchLastCommittedOffset(ctx)
+		if err == nil {
+			return offset, nil
+		}
+
+		level.Warn(r.logger).Log("msg", "failed to fetch last committed offset", "err", err)
+		retry.Wait()
+	}
+
+	// Handle the case the context was canceled before the first attempt.
+	if err == nil {
+		err = retry.Err()
+	}
+
+	return offset, err
+}
+
+func (r *PartitionReader) fetchLastCommittedOffset(ctx context.Context) (kgo.Offset, error) {
+	offsets, err := r.adm.FetchOffsets(ctx, r.consumerGroup)
+	if errors.Is(err, kerr.UnknownTopicOrPartition) {
+		// In case we are booting up for the first time ever against this topic.
+		return kgo.NewOffset().AtStart(), nil
+	}
+	if err != nil {
+		return kgo.NewOffset(), errors.Wrap(err, "unable to fetch group offsets")
+	}
+	offset, found := offsets.Lookup(r.topic, r.partitionID)
+	if !found {
+		// No committed offset found for this partition, start from the end
+		return kgo.NewOffset().AtStart(), nil
+	}
+	return kgo.NewOffset().At(offset.At), nil
+}
+
+func (r *PartitionReader) commitLoop(ctx context.Context) {
+	defer r.wg.Done()
+
+	ticker := time.NewTicker(r.commitInterval)
+	defer ticker.Stop()
+
+	lastCommittedOffset := int64(-1)
+
+	for {
+		select {
+		case <-ctx.Done():
+			// Final commit on shutdown
+			r.commitCurrentWatermark(lastCommittedOffset)
+			return
+		case <-ticker.C:
+			// Periodic commit
+			lastCommittedOffset = r.commitCurrentWatermark(lastCommittedOffset)
```

Comment:
How do the offsets interact? Since we only sync the watermark every X time, that there is the possibility for replay in regards for consuming more records. But we dont use the offset in the records structure in the live store consume function and there doesnt appear to be an logic to check for duplicate entries?

Reply:
Added a comment mentioning the tradeoffs of this approach